### PR TITLE
Yet Another Composable Stable Pool Deployment

### PR DESCRIPTION
**Pull request by TomAFrench**
# Description

#2053 but with extra time travel.

## Type of change

- [ ] Bug fix 
- [ ] New feature 
- [ ] Breaking change 
- [ ] Dependency changes
- [ ] Code refactor / cleanup
- [ ] Documentation or wording changes
- [x] Other

## Checklist:

- [x] The diff is legible and has no extraneous changes
- [x] Complex code has been commented, including external interfaces
- [x] Tests are included for all code paths
- [x] The base branch is either `master`, or there's a description of how to merge

### CODE REVIEW

File: pkg/pool-stable/contracts/ComposableStablePoolFactory.sol [hunk 1]
```diff
@@ -23,13 +25,28 @@ import "@balancer-labs/v2-pool-utils/contracts/factories/FactoryWidePauseWindow.
 
 import "./ComposableStablePool.sol";
 
-contract ComposableStablePoolFactory is BasePoolSplitCodeFactory, FactoryWidePauseWindow {
+contract ComposableStablePoolFactory is IVersion, IPoolVersion, BasePoolSplitCodeFactory, FactoryWidePauseWindow {
```
Comment (jubeira) [inline pkg/pool-stable/contracts/ComposableStablePoolFactory.sol:28]: `BasePoolSplitCodeFactory` or `BasePoolFactory`?

File: pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol [hunk 2]
```diff
@@ -75,7 +78,7 @@ abstract contract ComposableStablePoolProtocolFees is
         // swap fees and yield.
         (
             uint256 expectedProtocolOwnershipPercentage,
-            uint256 totalGrowthInvariant
+            uint256 currentInvariantWithLastJoinExitAmp
         ) = _getProtocolPoolOwnershipPercentage(balances, lastJoinExitAmp, lastPostJoinExitInvariant);
 
         // Now that we know what percentage of the Pool's current value the protocol should own, we can compute how
```
Comment (jubeira) [inline pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol:84]: Same below this comment:

```
        uint256 protocolFeeAmount = ProtocolFees.bptForPoolOwnershipPercentage(
            virtualSupply,
            expectedProtocolOwnershipPercentage
        );
```

File: pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol [hunk 3]
```diff
@@ -279,21 +290,28 @@ abstract contract ComposableStablePoolProtocolFees is
...
+            // To convert to a percentage of pool ownership, multiply by the rate,
+            // then normalize against the final invariant
+            uint256 protocolOwnershipPercentage = Math.divDown(
+                Math.mul(invariantDeltaFromFees, getProtocolFeePercentageCache(ProtocolFeeType.SWAP)),
+                postJoinExitInvariant
             );
 
-            _payProtocolFees(protocolFeeAmount);
+            if (protocolOwnershipPercentage > 0) {
+                uint256 protocolFeeAmount = _calculateAdjustedProtocolFeeAmount(
```
Comment (jubeira) [inline pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol:308]: Same here:

```
                uint256 protocolFeeAmount = ProtocolFees.bptForPoolOwnershipPercentage(
                    postJoinExitSupply,
                    protocolOwnershipPercentage
                );
```

File: pkg/pool-stable/contracts/ComposableStablePoolRates.sol [hunk 4]
```diff
@@ -186,21 +186,9 @@ abstract contract ComposableStablePoolRates is ComposableStablePoolStorage {
...
-        _cacheTokenRateIfNecessary(3);
-
-        if (totalTokens == 4) return;
-        _cacheTokenRateIfNecessary(4);
-
-        if (totalTokens == 5) return;
-        _cacheTokenRateIfNecessary(5);
+        for (uint256 i = 0; i < totalTokens; ++i) {
+            _cacheTokenRateIfNecessary(i);
+        }
```
Comment (jubeira) [inline pkg/pool-stable/contracts/ComposableStablePoolRates.sol:191]: Note: These are not in the diff in #2059, but were present in the base already (applies to all these refactors in `ComposableStablePoolRates`.

File: pkg/pool-stable/contracts/StableMath.sol [hunk 5]
```diff
@@ -448,14 +448,28 @@ library StableMath {
...
+
+        // Tokens in, so we round up overall.
+        uint256 bptRatio = bptAmountOut.divUp(totalBPT);
+
+        uint256[] memory amountsIn = new uint256[](balances.length);
+        for (uint256 i = 0; i < balances.length; i++) {
+            amountsIn[i] = balances[i].mulUp(bptRatio);
+        }
+
+        return amountsIn;
```
Comment (jubeira) [inline pkg/pool-stable/contracts/StableMath.sol:473]: Note: These are not in the diff in https://github.com/balancer-labs/balancer-v2-monorepo/pull/2059, but were present in the base already.

File: pkg/pool-stable/contracts/StablePoolAmplification.sol [hunk 6]
```diff
@@ -193,6 +193,7 @@ abstract contract StablePoolAmplification is BasePoolAuthorization {
     function _isOwnerOnlyAction(bytes32 actionId) internal view virtual override returns (bool) {
         return
             (actionId == getActionId(this.startAmplificationParameterUpdate.selector)) ||
-            (actionId == getActionId(this.stopAmplificationParameterUpdate.selector));
+            (actionId == getActionId(this.stopAmplificationParameterUpdate.selector)) ||
+            super._isOwnerOnlyAction(actionId);
```
Comment (jubeira) [inline pkg/pool-stable/contracts/StablePoolAmplification.sol:197]: Note: These are not in the diff in https://github.com/balancer-labs/balancer-v2-monorepo/pull/2059, but were present in the base already.

File: pkg/pool-stable/test/ComposableStablePool.test.ts [hunk 7]
```diff
@@ -378,9 +390,12 @@ describe('ComposableStablePool', () => {
 
         it('returns the pre-join invariant', async () => {
           const { value, precision } = await pool.getAmplificationParameter();
+
+          // We pass a floating point amplification instead of an integer one, to more closely match the behavior in the
+          // Pool.
           const expectedInvariant = calculateInvariant(
             registeredBalances.filter((_, i) => i !== bptIndex),
-            value.div(precision)
+            value.toNumber() / precision.toNumber()
```
Comment (jubeira) [inline pkg/pool-stable/test/ComposableStablePool.test.ts:398]: Note: These are not in the diff in https://github.com/balancer-labs/balancer-v2-monorepo/pull/2059, but were present in the base already.

File: pkg/pool-stable/test/ComposableStablePool.test.ts [hunk 8]
```diff
@@ -1539,30 +1643,39 @@ describe('ComposableStablePool', () => {
       });
     });
 
-    describe('getRate', () => {
+    describe('getRate and protocol fees', () => {
       const swapFeePercentage = fp(0.1); // 10 %
       const protocolFeePercentage = fp(0.5); // 50 %
 
-      sharedBeforeEach('deploy pool', async () => {
+      sharedBeforeEach('deploy  pool', async () => {
```
Comment (jubeira) [inline pkg/pool-stable/test/ComposableStablePool.test.ts:1650]: nit
```suggestion
      sharedBeforeEach('deploy pool', async () => {
```

File: pkg/pool-utils/contracts/BasePool.sol [hunk 9]
```diff
@@ -264,7 +274,8 @@ abstract contract BasePool is IBasePool, BasePoolAuthorization, BalancerPoolToke
     function _isOwnerOnlyAction(bytes32 actionId) internal view virtual override returns (bool) {
         return
             (actionId == getActionId(this.setSwapFeePercentage.selector)) ||
-            (actionId == getActionId(this.setAssetManagerPoolConfig.selector));
+            (actionId == getActionId(this.setAssetManagerPoolConfig.selector)) ||
+            super._isOwnerOnlyAction(actionId);
```
Comment (jubeira) [inline pkg/pool-utils/contracts/BasePool.sol:278]: Note: all these changes were already included in the base of #2059.

The only difference is that at the base of #2059, `BasePool` implements `IControlledPool`. That seems to be just a refactor; the functionality is the same.

File: pkg/pool-utils/contracts/BasePoolAuthorization.sol [hunk 10]
```diff
@@ -55,7 +55,9 @@ abstract contract BasePoolAuthorization is Authentication {
         }
     }
 
-    function _isOwnerOnlyAction(bytes32 actionId) internal view virtual returns (bool);
+    function _isOwnerOnlyAction(bytes32) internal view virtual returns (bool) {
+        return false;
+    }
```
Comment (jubeira) [inline pkg/pool-utils/contracts/BasePoolAuthorization.sol:60]: Note: all these changes were already included in the base of https://github.com/balancer-labs/balancer-v2-monorepo/pull/2059.

File: pkg/pool-utils/contracts/ProtocolFeeCache.sol [hunk 11]
```diff
@@ -126,6 +128,13 @@ abstract contract ProtocolFeeCache is RecoveryMode {
         _updateProtocolFeeCache(_protocolFeeProvider, ProtocolFeeType.AUM);
     }
 
+    /**
+     * @dev Override in derived contracts to perform some action before the cache is updated. This is typically relevant
+     * to Pools that incur protocol debt between operations. To avoid altering the amount due retroactively, this debt
+     * needs to be paid before the fee percentages change.
+     */
+    function _beforeProtocolFeeCacheUpdate() internal virtual {}
+
```
Comment (jubeira) [inline pkg/pool-utils/contracts/ProtocolFeeCache.sol:137]: Note: all these changes were already included in the base of https://github.com/balancer-labs/balancer-v2-monorepo/pull/2059.

**Review by jubeira — COMMENTED** [review]
This is looking pretty good.

I went through #2059 all over again, checking not only the diff but the actual files right before merging with master @https://github.com/balancer-labs/balancer-v2-monorepo/tree/a49fe13ff97d881ded2a7f9c4bf5dff2578ee5ad and comparing that with this diff and how the files look at the top @https://github.com/balancer-labs/balancer-v2-monorepo/pull/2069/commits/80a0c67a784e25c353dceae1536e764b59a86f33.

I've found just a few differences to review:
- the use of `_calculateAdjustedProtocolFeeAmount` in `ComposableStablePool` vs using `ProtocolFees.bptForPoolOwnershipPercentage`. This seems to be a refactor that was part of the base of #2059 and was not included here; the functionality is the same, so the current version in this PR should be OK.
- In `ComposableStablePoolFactory`, #2059 inherits from `BasePoolFactory` instead of `BasePoolSplitCodeFactory`.
- In #2059, `BasePool` implements `IControlledPool`. This seems to be just a refactor included in the base without any impact in functionality.

Other than that, I didn't find

File: pkg/pool-stable/test/ComposableStablePool.test.ts (hunk 8 above)
Comment (jubeira) [reply pkg/pool-stable/test/ComposableStablePool.test.ts:1650]: (this can be addressed later on)

File: pkg/pool-stable/contracts/ComposableStablePoolFactory.sol (hunk 1 above)
Comment (nventuro) [reply pkg/pool-stable/contracts/ComposableStablePoolFactory.sol:28]: `BasePoolSplitCodeFactory` was eventually renamed to `BasePoolFactory`, which was deleted.

Comment (TomAFrench) [reply pkg/pool-stable/contracts/ComposableStablePoolFactory.sol:28]: I'm not sure on the question here. Why would we switch from the factory type that deployed the previous pool?

File: pkg/pool-stable/test/ComposableStablePool.test.ts (hunk 8 above)
Comment (TomAFrench) [reply pkg/pool-stable/test/ComposableStablePool.test.ts:1650]: I'd say this is out of scope for this PR. I'd want to keep the diff for this compared to the previous pool as small as possible.

File: pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol (hunk 2 above)
Comment (TomAFrench) [reply pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol:84]: Sorry could you rephrase?

File: pkg/pool-stable/contracts/ComposableStablePoolFactory.sol (hunk 1 above)
Comment (jubeira) [reply pkg/pool-stable/contracts/ComposableStablePoolFactory.sol:28]: I just wanted to double check if there was any difference between the two (`BasePoolSplitCodeFactory` or `BasePoolFactory`).

`BasePoolSplitCodeFactory` was present in the base, and `BasePoolFactory` is in master. I wasn't sure whether this was a change that we wanted to introduce as part of the changes between composable v1 and composable v2 or not, but if it's just a name change it should be the same.

File: pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol (hunk 2 above)
Comment (jubeira) [reply pkg/pool-stable/contracts/ComposableStablePoolProtocolFees.sol:84]: In [L84/L87](https://github.com/balancer-labs/balancer-v2-monorepo/pull/2069/files#diff-06011c15b516ce6b965c6ed91b1b6bc55af19e8ffa3cf37b08ebe07bbfb5b193L84) below this line (you'll need to expand down since that line wasn't changed), the file does 

```
        uint256 protocolFeeAmount = _calculateAdjustedProtocolFeeAmount(
            virtualSupply,
            expectedProtocolOwnershipPercentage
        );
```

This was at some point factored out to a library, like this:

```
        uint256 protocolFeeAmount = ProtocolFees.bptForPoolOwnershipPercentage(
            virtualSupply,
            expectedProtocolOwnershipPercentage
        );
```

The functionality should be the same in any case; it's just a function that was factored out to a library, but I wanted to raise it to double check.

**Comment by TomAFrench** [comment]
Closing in favour of #2065
